keps/sig-cli/2985-public-krm-functions-registry/README.md

```diff
@@ -0,0 +1,1344 @@
+                    description: If network is required to run this function.
+                    type: boolean
+                  requireVolumeMount:
+                    description: If volumn mount is required to run this function.
+                    type: boolean
+              exec:
+                description: The runtime information for exec-based KRM function.
+                type: object
+                required:
+                  - platform
+                properties:
+                  platforms:
+                    description: Per platform runtime information.
+                    type: array
+                    items:
+                      type: object
+                      required:
+                        - bin
+                        - os
+                        - arch
+                        - uri
+                        - sha256
+                      properties: 
+                        bin:
+                          description: The binary name.
+                          type: string
+                        os:
+                          description: The target operation system to run the KRM function.
+                          type: string
+                          enum:
+                            - linux
+                            - darwin
+                            - windows
+                        arch:
+                          description: The target archtechture to run the KRM function.
+                          type: string
+                          enum:
+                            - amd64
+                            - arm64
+                        uri:
+                          description: The location to download the binary.
+                          type: string
+                        sha256:
+                          description: The degist of the binary that can be used to verify the binary.
+                          type: string
+paths: {}
+```
+</details>
+
+### Publishing Workflow
+
+We only support publishing container-based KRM function in the public registry.
+We will only cover the workflow for that.
+
+The developer needs to do the following:
+
+1. Build a container image and pushed it to a publicly accessible container
+   registry.
+2. Ensure the usage doc is a markdown file and is up-to-date and publicly
+   accessible.
+3. Create a file called `krm-function-metadata.yaml` which contains the metadata
+   that satisfies the KRM function metadata schema above.
+4. Checkout the KRM function registry repo.
+5. Move the `krm-function-metadata.yaml` file to the desired location in the
+   repo by following the repo layout convention (discussed below).
+6. Depending on the requirements (discussed in an earlier session) of different
+   publisher type, choose the right email to commit the change.
+7. Create a PR and get it reviewed and approved by the publisher OWNERS.
+
+### Repo Layout Convention
+
+```shell
+├── publishers
+│   ├── communities
```

Comment:
why would we want to separate those?  What advantage does that give us?  Hypothetical question - why don't I just have a one long list fo functions at the root level by name and the owner metadata is already specified?

We now introduce a dependency between what's in the file and where it exists.  People might forget to move if for instance the publisher changes (gets acquired for instance).

Reply:
Different types of publishers will have slightly different requirements. When we saw a publisher Foo in a function metadata, we need to know the type of Foo (community, company or individual) to determine the verification requirement needed.
I'm open to suggestions about the layout.

>We now introduce a dependency between what's in the file and where it exists. People might forget to move if for instance the publisher changes (gets acquired for instance).

Are you worrying about that the image may disappear (e.g. renamed or moved) after publishing the function?
To mitigate that, we can have some cron-based CI like checking nightly to see if the images still exist. If not, notify the maintainers by GH issue and(or) email.